src/main/java/com/atlassian/db/replica/internal/ConnectionParameters.java

```diff
@@ -14,8 +12,13 @@
     private String catalog;
     private Map<String, Class<?>> typeMap;
     private Integer holdability;
+    private final Set<Connection> initializedConnections = Collections.newSetFromMap(new WeakHashMap<>());
```

Comment:
There's `State.NOT_INITIALISED`. Is that the same thing? Also `ConnectionState` has calls like `readConnection.isInitialized()`. Is that related?

It starts to be hard to think about. `DualConnection` has some immutable **state**. `ReplicaConnectionProvider` has the  `ConnectionParameters` *and* `ConnectionState` as its state. Then `ConnectionState` has the `ConnectionParameters` *again* and it returns a different `State`.

It seems we lost the OOP way and went back to data structures and procedures. This grows the code, because reuse is limited, because different procedures work on different data structures rather than on the same objects.

Reply:
I think we can still go a little further in this style to get replicas running on prod. But soon after, the tech debt will need to be paid off.

Reply:
> There's State.NOT_INITIALISED. Is that the same thing?

No. It's the state of the dual connection. It means no connection was initialized. We use real connections internally. We may initialise them multiple times (first replica, then main connection, because replica was inconsistent, and again replica as it's already consistent ....). Both are states but are not related.

`readConnection.isInitialized()` it's from a lazy connection. Now it means only it's obtained from the pool. Maybe it's a good idea to initialize the parameters here and return only fully-initialized connections from the pool.

Reply:
@dagguh, I refactored it. Is it better now?

Reply:
> It seems we lost the OOP way and went back to data structures and procedures.

Can you elaborate?

Reply:
> @dagguh, I refactored it. Is it better now?

Much simpler, whoa! My 🐒🧠 can track it now 😉

Reply:
In OOP, objects protect their state via encapsulation and protect the control over them by exposing only behaviors (not commands like setters). For example, if `ConnectionState` is a state of a connection, it should be encapsulated in the connection. It also steals behaviors from `State` and `Connection` by issuing commands to them and by digging in their internals.

Reply:
`State` has no behaviours. It's an enum. `ConnectionState` exposes the current `state` of DualConnection. To obtain the current state, it needs to verify what raw `connections` are initialized (as the `state` describes it). It's also the only point to change the state as it encapsulates the raw connections. It would lose control over changing and retrieving the state without it. Currently, ConnectionState can:
- control the state --> "issuing commands on raw connections."
- retrieve the state--> "steals behaviors from State"

`ConnectionState` is encapsulated in `DualConnection` (unfortunately, there's also wrongly named mediator `ReplicaConnectionProvider`, but it's something to refactor in the future). 

I still don't get what's not OO here. Maybe you mix DualConnection (smart proxy over real connections) with raw connections (state of the DualConnection).

Reply:
We could extract control over the state from `ConnectionState` to a different class. Would it make it more OO?

(there are also some methods that seem to be unrelated to the ConnectinState that we can remove `hasWriteConnection`, `getDecision`)

Reply:
It would reduce ConnectionState to 

```
 public State getState() {
        if (isClosed != null && isClosed) {
            return CLOSED;
        } else {
            final boolean readReady = readConnection.isInitialized();
            final boolean writeReady = writeConnection.isInitialized();
            if (!readReady && !writeReady) {
                return NOT_INITIALISED;
            } else if (!replicaConsistent && writeReady) {
                return COMMITED_MAIN;
            } else if (writeReady) {
                return MAIN;
            } else {
                return REPLICA;
            }
        }
    }
```

Reply:
These "state" classes look like a bigger object was cut into pieces, but not with a clean edge. It's like they still have leftover connections between them and they expose fields and methods to each other to keep working. By doing that, they leak it to others too. Objects have behaviors and protect their internals, so enums and DTOs are not objects (all state no behavior).

Reply:
> It would reduce ConnectionState to

Yeah, reducing or entirely dissolving a "state" object would definitely help encapsulation

Reply:
> These "state" classes look like a bigger object was cut into pieces, but not with a clean edge.

Yes, ConnectionState cut it from multiple places. It's not the final shape it should be. 
So I think we have a consensus to fix it by extracting part of the ConnectionState to more appropriate places.
But I don't get this part:

> dissolving a "state" object

Can you elaborate? We didn't have an explicit `state` before, so all the checks were scattered here and there. State reduces it to one place. You no longer decide how to act depending on random ifs that checks what connection is initialized, it's all unified in one place.

Reply:
I plan to use some of my 20% for refactoring. As you mentioned, let's make it green and then we can continue. For me, the priorities are:
- make it work 😉 ,
- introduce a stable API (1.0.0 release), 
- refactor internals.

Let us start with the refactor that may affect API.

@dagguh Thanks for all the feedback and for looking at the code with fresh eyes. It went a long way from the in-product spike to the lib. It went from ultra-complex code to less complex, but still, it requires improvement. I'd appreciate it if you start to refactor to show proposed changes. It would be much easier to talk about the changes.

Reply:
Gathering control from all around the codebase into a few objects was indeed a good and needed move. It remained procedural, because "explicit state" is a data structure. And with objects we want to hide them, exposing only behaviors. I'll open a refactor PR next week 👍